pkg/services/hcloud/server/server.go

```diff
@@ -426,57 +893,97 @@ func (s *Service) reconcileLoadBalancerAttachment(ctx context.Context, server *h
+func (s *Service) createServerFromImageNameOrURL(ctx context.Context) (*hcloud.Server, *hcloud.Image, error) {
+	if s.scope.HCloudMachine.Spec.ImageName != "" {
+		return s.createServerFromImageName(ctx)
+	}
+	return s.createServerFromImageURL(ctx)
+}
+
+func (s *Service) createServerFromImageURL(ctx context.Context) (*hcloud.Server, *hcloud.Image, error) {
+	// Validate that ImageURLCommand is given
+	hm := s.scope.HCloudMachine
+	if s.scope.ImageURLCommand == "" {
+		msg := fmt.Sprintf("controller was started without --hcloud-image-url-command. Provisioning via ImageURL %q not possible", hm.Spec.ImageURL)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return nil, nil, errors.New(msg)
+	}
+
+	image, err := s.getServerImage(ctx, preRescueOSImage)
+	if err != nil {
+		return nil, nil, fmt.Errorf("createServerFromImageURL: failed to get server image: %w", err)
+	}
+	server, err := s.createServer(ctx, s.scope.Name(), nil, image)
+	if err != nil {
+		return nil, nil, err
+	}
+	s.scope.HCloudMachine.SetBootState(infrav1.HCloudBootStateWaitForPreRescueOSThenEnableRescueSystem)
+	return server, image, nil
+}
+
 func (s *Service) createServerFromImageName(ctx context.Context) (*hcloud.Server, *hcloud.Image, error) {
+	hm := s.scope.HCloudMachine
 	userData, err := s.scope.GetRawBootstrapData(ctx)
 	if err != nil {
-		record.Warnf(
-			s.scope.HCloudMachine,
-			"FailedGetBootstrapData",
-			err.Error(),
-		)
-		return nil, nil, fmt.Errorf("failed to get raw bootstrap data: %s", err)
+		err = fmt.Errorf("failed to get raw bootstrap data: %s", err)
+		msg := err.Error()
+		record.Warn(hm, "FailedGetBootstrapData", msg)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", err.Error())
+		return nil, nil, err
 	}
-	image, err := s.getServerImage(ctx)
+
+	image, err := s.getServerImage(ctx, hm.Spec.ImageName)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to get server image: %w", err)
+		err = fmt.Errorf("failed to get server image: %w", err)
+		msg := err.Error()
+		record.Warn(hm, "FailedGetBootstrapData", msg)
+		conditions.MarkFalse(hm, infrav1.ServerAvailableCondition,
+			string(hm.Status.BootState), clusterv1.ConditionSeverityWarning,
+			"%s", msg)
+		return nil, nil, err
 	}
 
-	server, err := s.createServer(ctx, userData, image)
+	server, err := s.createServer(ctx, s.scope.Name(), userData, image)
 	if err != nil {
 		return nil, nil, err
 	}
 
-	s.scope.HCloudMachine.SetBootState(infrav1.HCloudBootStateBootToRealOS)
+	hm.SetBootState(infrav1.HCloudBootStateBootToRealOS)
 	return server, image, nil
 }
 
-func (s *Service) createServer(ctx context.Context, userData []byte, image *hcloud.Image) (*hcloud.Server, error) {
+func (s *Service) createServer(ctx context.Context, serverName string, userData []byte, image *hcloud.Image) (*hcloud.Server, error) {
```

Comment:
I think the changes to this function are irrelevant. Why do you add the server name in this way?